File: app/routers/holdings.py
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.external.market_data import RateLimitError, get_quote
from app.models import Account, Holding
from app.schemas.account import (
    HoldingCreate,
    HoldingResponse,
    HoldingUpdate,
    RefreshPricesRequest,
    RefreshPricesResponse,
)
from app.services.core import holding_to_response, remove_holding
from app.utils import (
    ASSET_CLASS_DEPOSIT,
    generate_deposit_symbol,
    is_deposit_holding,
    normalize_deposit_holding,
)

router = APIRouter(prefix="/holdings", tags=["holdings"])


def _apply_market_quote(holding: Holding) -> None:
    if is_deposit_holding(holding):
        return
    try:
        quote = get_quote(holding.symbol)
    except (ValueError, RateLimitError):
        return
    holding.last_market_price = quote.price
    holding.last_price_updated_at = datetime.utcnow()
    if quote.name:
        holding.name = quote.name


def _prepare_holding_fields(payload: HoldingCreate) -> dict:
    asset_class = payload.asset_class
    if asset_class == ASSET_CLASS_DEPOSIT:
        return {
            "asset_class": asset_class,
            "symbol": payload.symbol or generate_deposit_symbol(),
            "name": payload.name,
            "quantity": payload.quantity,
            "avg_cost_price": payload.avg_cost_price,
            "manual_price": payload.avg_cost_price,
            "interest_rate": payload.interest_rate,
            "start_date": payload.start_date or date.today(),
            "maturity_date": payload.maturity_date,
        }
    return {
        "asset_class": asset_class,
        "symbol": payload.symbol,
        "name": payload.name,
        "quantity": payload.quantity,
        "avg_cost_price": payload.avg_cost_price,
        "manual_price": None,
        "interest_rate": None,
        "start_date": None,
        "maturity_date": None,
    }


@router.get("", response_model=list[HoldingResponse])
def list_holdings(account_id: int | None = None, db: Session = Depends(get_db)):
    query = db.query(Holding).options(joinedload(Holding.account)).filter(Holding.quantity > 0)
    if account_id:
        query = query.filter(Holding.account_id == account_id)
    holdings = query.order_by(Holding.id).all()
    return [holding_to_response(h, h.account.name if h.account else None) for h in holdings]


@router.post("", response_model=HoldingResponse, status_code=201)
def create_holding(payload: HoldingCreate, db: Session = Depends(get_db)):
    account = db.get(Account, payload.account_id)
    if not account:
        raise HTTPException(status_code=404, detail="계좌를 찾을 수 없습니다.")
    if not account.account_type.supports_holdings:
        raise HTTPException(status_code=400, detail="이 계좌 유형은 보유 종목을 지원하지 않습니다.")

    fields = _prepare_holding_fields(payload)
    holding = Holding(account_id=payload.account_id, **fields)
    if is_deposit_holding(holding):
        normalize_deposit_holding(holding)
    db.add(holding)
    db.flush()
    _apply_market_quote(holding)
    db.commit()
    db.refresh(holding)
    holding = db.query(Holding).options(joinedload(Holding.account)).filter(Holding.id == holding.id).first()
    return holding_to_response(holding, holding.account.name)


@router.patch("/{holding_id}", response_model=HoldingResponse)
def update_holding(holding_id: int, payload: HoldingUpdate, db: Session = Depends(get_db)):
    holding = db.query(Holding).options(joinedload(Holding.account)).filter(Holding.id == holding_id).first()
    if not holding:
        raise HTTPException(status_code=404, detail="보유 종목을 찾을 수 없습니다.")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(holding, key, value)
    if is_deposit_holding(holding):
        normalize_deposit_holding(holding)
    db.commit()
    db.refresh(holding)
    return holding_to_response(holding, holding.account.name)


@router.delete("/{holding_id}", status_code=204)
def delete_holding(holding_id: int, db: Session = Depends(get_db)):
    holding = db.get(Holding, holding_id)
    if not holding:
        raise HTTPException(status_code=404, detail="보유 종목을 찾을 수 없습니다.")
    remove_holding(db, holding)
    db.commit()


@router.post("/refresh-prices", response_model=RefreshPricesResponse)
def refresh_prices(payload: RefreshPricesRequest | None = None, db: Session = Depends(get_db)):
    query = db.query(Holding).filter(Holding.quantity > 0, Holding.asset_class != ASSET_CLASS_DEPOSIT)
    if payload and payload.holding_ids:
        query = query.filter(Holding.id.in_(payload.holding_ids))
    holdings = query.all()
    updated = 0
    failed: list[dict] = []
    for holding in holdings:
        try:
            quote = get_quote(holding.symbol)
            holding.last_market_price = quote.price
            holding.last_price_updated_at = datetime.utcnow()
            if quote.name:
                holding.name = quote.name
            updated += 1
        except RateLimitError:
            failed.append(
                {"holding_id": holding.id, "symbol": holding.symbol, "reason": "시세 서버 요청 한도 초과"}
            )
            break
        except ValueError:
            failed.append(
                {"holding_id": holding.id, "symbol": holding.symbol, "reason": "시세 조회 실패"}
            )
    db.commit()
    return RefreshPricesResponse(updated=updated, failed=failed)

File: src/lib/services/holdings.ts
import type { Prisma } from "@prisma/client";
import Decimal from "decimal.js";

import { roundAvgCostPrice } from "@/lib/decimal";
import { prisma } from "@/lib/db";
import { getQuote, RateLimitError } from "@/lib/external/market-data";
import { ServiceError } from "@/lib/service-error";
import { removeHolding } from "@/lib/services/core";
import {
  ASSET_CLASS_DEPOSIT,
  generateDepositSymbol,
  isDepositHolding,
  normalizeDepositHolding,
} from "@/lib/utils";
import type { holdingCreateSchema } from "@/lib/validations/account";
import type { z } from "zod";

type HoldingCreate = z.infer<typeof holdingCreateSchema>;

export async function applyMarketQuote(holding: {
  asset_class?: string;
  symbol: string;
  name: string;
  quantity: Decimal.Value;
  last_market_price?: Decimal | null;
  last_price_updated_at?: Date | null;
}): Promise<void> {
  if (isDepositHolding(holding)) {
    return;
  }
  try {
    const quote = await getQuote(holding.symbol);
    holding.last_market_price = quote.price;
    holding.last_price_updated_at = new Date();
    if (quote.name) {
      holding.name = quote.name;
    }
  } catch {
    // silent swallow (matches Python ValueError/RateLimitError on create)
  }
}

export function prepareHoldingFields(payload: HoldingCreate) {
  if (payload.asset_class === ASSET_CLASS_DEPOSIT) {
    return {
      asset_class: payload.asset_class,
      symbol: payload.symbol ?? generateDepositSymbol(),
      name: payload.name,
      quantity: payload.quantity,
      avg_cost_price: roundAvgCostPrice(payload.avg_cost_price ?? 1),
      manual_price: payload.avg_cost_price ?? new Decimal(1),
      interest_rate: payload.interest_rate ?? null,
      start_date: payload.start_date ?? new Date(),
      maturity_date: payload.maturity_date ?? null,
    };
  }
  return {
    asset_class: payload.asset_class,
    symbol: payload.symbol!,
    name: payload.name,
    quantity: payload.quantity,
    avg_cost_price: roundAvgCostPrice(payload.avg_cost_price!),
    book_cost: payload.book_cost ?? null,
    manual_price: null,
    interest_rate: null,
    start_date: null,
    maturity_date: null,
  };
}

export async function refreshPrices(holdingIds?: number[] | null) {
  const holdings = await prisma.holding.findMany({
    where: {
      quantity: { gt: 0 },
      asset_class: { not: ASSET_CLASS_DEPOSIT },
      ...(holdingIds?.length ? { id: { in: holdingIds } } : {}),
    },
  });

  let updated = 0;
  const failed: Array<{ holding_id: number; symbol: string; reason: string }> = [];

  for (const holding of holdings) {
    try {
      const quote = await getQuote(holding.symbol);
      await prisma.holding.update({
        where: { id: holding.id },
        data: {
          last_market_price: quote.price,
          last_price_updated_at: new Date(),
          ...(quote.name ? { name: quote.name } : {}),
        },
      });
      updated += 1;
    } catch (error) {
      if (error instanceof RateLimitError) {
        failed.push({
          holding_id: holding.id,
          symbol: holding.symbol,
          reason: "시세 서버 요청 한도 초과",
        });
        break;
      }
      failed.push({
        holding_id: holding.id,
        symbol: holding.symbol,
        reason: "시세 조회 실패",
      });
    }
  }

  return { updated, failed };
}

export async function createHolding(payload: HoldingCreate) {
  const account = await prisma.account.findUnique({
    where: { id: payload.account_id },
    include: { account_type: true },
  });
  if (!account) {
    throw new ServiceError(404, "계좌를 찾을 수 없습니다.");
  }
  if (!account.account_type.supports_holdings) {
    throw new ServiceError(400, "이 계좌 유형은 보유 종목을 지원하지 않습니다.");
  }

  const fields = prepareHoldingFields(payload);
  const holding = await prisma.holding.create({
    data: { account_id: payload.account_id, ...fields },
    include: { account: true },
  });

  if (isDepositHolding(holding)) {
    normalizeDepositHolding(holding);
  }
  await applyMarketQuote(holding);

  return prisma.holding.update({
    where: { id: holding.id },
    data: {
      last_market_price: holding.last_market_price,
      last_price_updated_at: holding.last_price_updated_at,
      name: holding.name,
      avg_cost_price: holding.avg_cost_price,
      manual_price: holding.manual_price,
    },
    include: { account: true },
  });
}

export async function updateHolding(
  holdingId: number,
  data: Prisma.HoldingUpdateInput,
) {
  const existing = await prisma.holding.findUnique({
    where: { id: holdingId },
    include: { account: true },
  });
  if (!existing) {
    throw new ServiceError(404, "보유 종목을 찾을 수 없습니다.");
  }

  const updated = await prisma.holding.update({
    where: { id: holdingId },
    data,
    include: { account: true },
  });

  if (isDepositHolding(updated)) {
    normalizeDepositHolding(updated);
    return prisma.holding.update({
      where: { id: holdingId },
      data: {
        avg_cost_price: updated.avg_cost_price,
        manual_price: updated.manual_price,
      },
      include: { account: true },
    });
  }

  return updated;
}

export async function deleteHolding(holdingId: number) {
  const holding = await prisma.holding.findUnique({ where: { id: holdingId } });
  if (!holding) {
    throw new ServiceError(404, "보유 종목을 찾을 수 없습니다.");
  }
  await removeHolding(prisma, holding);
}
